flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/join/HashJoinOperator.java

```diff
@@ -0,0 +1,420 @@
+			this.table.free();
+			this.table = null;
+		}
+	}
+
+	public static HashJoinOperator newHashJoinOperator(
+			long minMemorySize,
+			long maxMemorySize,
+			long eachRequestMemorySize,
+			HashJoinType type,
+			GeneratedJoinCondition condFuncCode,
+			boolean reverseJoinFunction,
+			boolean[] filterNullKeys,
+			GeneratedProjection buildProjectionCode,
+			GeneratedProjection probeProjectionCode,
+			boolean tryDistinctBuildRow,
+			int buildRowSize,
+			long buildRowCount,
+			long probeRowCount,
+			RowType keyType) {
+		HashJoinParameter parameter = new HashJoinParameter(minMemorySize, maxMemorySize, eachRequestMemorySize,
+				type, condFuncCode, reverseJoinFunction, filterNullKeys, buildProjectionCode, probeProjectionCode,
+				tryDistinctBuildRow, buildRowSize, buildRowCount, probeRowCount, keyType);
+		switch (type) {
+			case INNER:
+				return new InnerHashJoinOperator(parameter);
+			case BUILD_OUTER:
+				return new BuildOuterHashJoinOperator(parameter);
+			case PROBE_OUTER:
+				return new ProbeOuterHashJoinOperator(parameter);
+			case FULL_OUTER:
+				return new FullOuterHashJoinOperator(parameter);
+			case SEMI:
+				return new SemiHashJoinOperator(parameter);
+			case ANTI:
+				return new AntiHashJoinOperator(parameter);
+			case BUILD_LEFT_SEMI:
+			case BUILD_LEFT_ANTI:
+				return new BuildLeftSemiOrAntiHashJoinOperator(parameter);
+			default:
+				throw new IllegalArgumentException("invalid: " + type);
+		}
+	}
+
+	static class HashJoinParameter implements Serializable {
+		long reservedMemorySize;
+		long maxMemorySize;
+		long perRequestMemorySize;
+		HashJoinType type;
+		GeneratedJoinCondition condFuncCode;
+		boolean reverseJoinFunction;
+		boolean[] filterNullKeys;
+		GeneratedProjection buildProjectionCode;
+		GeneratedProjection probeProjectionCode;
+		boolean tryDistinctBuildRow;
+		int buildRowSize;
+		long buildRowCount;
+		long probeRowCount;
+		RowType keyType;
+
+		HashJoinParameter(
+				long reservedMemorySize, long maxMemorySize, long perRequestMemorySize, HashJoinType type,
+				GeneratedJoinCondition condFuncCode, boolean reverseJoinFunction,
+				boolean[] filterNullKeys,
+				GeneratedProjection buildProjectionCode,
+				GeneratedProjection probeProjectionCode, boolean tryDistinctBuildRow,
+				int buildRowSize, long buildRowCount, long probeRowCount, RowType keyType) {
+			this.reservedMemorySize = reservedMemorySize;
+			this.maxMemorySize = maxMemorySize;
+			this.perRequestMemorySize = perRequestMemorySize;
+			this.type = type;
+			this.condFuncCode = condFuncCode;
+			this.reverseJoinFunction = reverseJoinFunction;
+			this.filterNullKeys = filterNullKeys;
+			this.buildProjectionCode = buildProjectionCode;
+			this.probeProjectionCode = probeProjectionCode;
+			this.tryDistinctBuildRow = tryDistinctBuildRow;
+			this.buildRowSize = buildRowSize;
+			this.buildRowCount = buildRowCount;
+			this.probeRowCount = probeRowCount;
+			this.keyType = keyType;
+		}
+	}
+
+	/**
+	 * Inner join.
```

Comment:
Add some comments to each operator to explain each one's behavior. This comment is even less informative than the class name...